src/modelfree/shaping_wrappers.py

```diff
@@ -0,0 +1,215 @@
+        return self.venv.reset()
+
+    def step_wait(self):
+        obs, rew, done, infos = self.venv.step_wait()
+        for env_num in range(self.num_envs):
+            shaped_reward = 0
+            for rew_term, rew_value in infos[env_num][self.agent_idx].items():
+                # our shaping_params dictionary is hierarchically separated the rew_terms
+                # into two rew_types, 'dense' and 'sparse'
+                if rew_term in self.shaping_params['sparse']:
+                    rew_type = 'sparse'
+                elif rew_term in self.shaping_params['dense']:
+                    rew_type = 'dense'
+                else:
+                    continue
+
+                # weighted_reward := our weight for that term * value of that term
+                weighted_reward = self.shaping_params[rew_type][rew_term] * rew_value
+                self.step_rew_dict[rew_type][env_num].append(weighted_reward)
+
+                # perform annealing if necessary and then accumulate total shaped_reward
+                if self.reward_annealer is not None:
+                    c = self.get_annealed_exploration_reward()
+                    if rew_type == 'sparse':
+                        weighted_reward *= (1 - c)
+                    else:
+                        weighted_reward *= c
+                shaped_reward += weighted_reward
+
+            # log the results of an episode into buffers and then pass on the shaped reward
+            if done[env_num]:
+                for rew_type in self.step_rew_dict:
+                    self.ep_rew_dict[rew_type].append(sum(self.step_rew_dict[rew_type][env_num]))
+                    self.step_rew_dict[rew_type][env_num] = []
+            self.counter += 1
+            self._log_sparse_dense_rewards()
+            rew[env_num] = shaped_reward
+
+        return obs, rew, done, infos
+
+    def get_annealed_exploration_reward(self):
+        """
+        Returns c (which will be annealed to zero) s.t. the total reward equals
+        c * reward_move + (1 - c) * reward_remaining
+        """
+        c = self.reward_annealer()
+        assert 0 <= c <= 1
+        return c
+
+
+class NoisyAgentWrapper(BaseRLModel):
+    def __init__(self, agent, noise_annealer, noise_type='gaussian'):
+        """
+        Wrap an agent and add noise to its actions
+        :param agent: BaseRLModel the agent to wrap
+        :param noise_annealer: Annealer.get_value - presumably the noise should be decreased
+        over time in order to get the adversarial policy to perform well on a normal victim.
+        :param noise_type: str - the type of noise parametrized by noise_annealer's value.
+        Current options are [gaussian]
+        """
+        self.agent = agent
+        self.sess = agent.sess
+        self.noise_annealer = noise_annealer
+        self.noise_generator = self._get_noise_generator(noise_type)
+
+    @classmethod
+    def get_noise_params(cls, noise_params, env_name):
```

Comment:
Deja-vu with `get_shaping_params`. Refactor this `json` loading code out to a separate function to code down on code duplication.